### ISSUE TITLE
Implement `Try` and `FromResidual<!>` for `()`

### ISSUE THREAD

**Issue by Jarcho** [rust-lang/libs-team#187]
# Proposal

## Problem statement

Using `ControlFlow<Self::T>` in a trait where most implementors set `T = !`. (or `Result<_, Self::T>`) results in a lot of boilerplate for the implementer.

## Motivation, use-cases

Visitors with an early return can be modeled as a trait such as the following (used in rustc):
```rust
trait Visitor {
    type BreakTy = !;
    fn visit_foo(&mut self, foo: &Foo) -> ControlFlow<Self::BreakTy> {
        walk_foo(self, foo)
    }
    // And more `visit_*` functions
}

fn walk_foo<V: Visitor>(v: &V, foo: &Foo) -> ControlFlow<V::BreakTy> {
    // walk foo using `?` when needed
    Continue(())
}
```

With non-branching implementations looking like (actual implementations can get worse):

```rust
impl Visitor for Bar {
    fn visit_foo(&mut self, foo: &Foo) -> ControlFlow<!> {
        if some_condition() {
            // do something
        } else if something_else() {
            // do something
        } else {
            walk_foo();
        }
        Continue(())
    }
}
```

## Solution sketches

If `Try` and `FromResidual<!>` are implemented for `()`, then the previous example would instead be:
```rust
trait Visitor {
    type ResultTy: Try<Output = ()> = ();
    fn visit_foo(&mut self, foo: &Foo) -> Self::ResultTy {
        walk_foo(self, foo)
    }
    // And more `visit_*` functions
}

fn walk_foo<V: Visitor>(v: &V, foo: &Foo) -> V::ResultTy {
    // walk foo using `?` when needed
    V::ResultTy::from_output(())
}

impl Visitor for Bar {
    fn visit_foo(&mut self, foo: &Foo) {
        if some_condition() {
            // do something
        } else if something_else() {
            // do something
        } else {
            walk_foo();
        }
    }
}
```

Any impl which requires branching can then set `ResultTy` to `ControlFlow<T>` (or `Result<(), T>` if that makes more sense).

---

An alternative would be to use a custom `ResultTy` trait implemented for both `()` and `ControlFlow<T>` (and others if needed). This would allow effectively the same interface, except anything which interacts with generic `Visitor`s wouldn't be able to use the `?` operator. This would mainly affect the implementation of `walk_*` functions which would otherwise make heavy use of it.

Downsides would be allowing `foo()?????` when `foo` returns `()`. This is easily solved with a lint any any use of `?` on an expression of the type `()`.

## Links and related work

Came up on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Use.20.60ControlFlow.60.20in.20HIR.20.60Visitor.60.20compiler-team.23597) while discussing using `ControlFlow` more.

## What happens now?

This issue is part of the libs-api team [API change proposal process]. Once this issue is filed the libs-api team will review open proposals in its weekly meeting. You should receive feedback within a week or two.

[API change proposal process]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

**Comment by scottmcm** [comment]
(Since I was mentioned in the meeting notes) I could go either way on this.  See my previous comments in <https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Use.20.60ControlFlow.60.20in.20HIR.20.60Visitor.60.20compiler-team.23597/near/339419883>.

I feel kinda weird about it, like maybe it wants something different, but it also meets all the expected properties, and `()?` could be linted about strongly, like how we're about to start linking on `let x = returns_unit();`, so I don't think it's necessarily unreasonable either.

TLDR I'm happy to follow libs-api's instincts either way on this one 🙂

**Comment by Amanieu** [comment]
Closing ACP as per the above comment.
